nntrainer/tensor/tensor_pool.cpp

```diff
@@ -0,0 +1,214 @@
+ * @note we assume that the caller checks if the exec_order and lifespan are
+ * compatible.
+ */
+Tensor *TensorPool::requestTensor(const TensorDim dim,
+                                  const std::vector<unsigned int> &exec_order,
+                                  TensorLifespan lifespan,
+                                  const std::string &name) {
+  if (pool.find(name) != pool.end())
+    throw std::invalid_argument("Cannot request tensor with same name");
+
+  if (dim.getDataLen() == 0)
+    throw std::invalid_argument("Cannot request tensor with size 0");
+
+  if (name.empty())
+    throw std::invalid_argument("Cannot request tensor with empty name");
+
+  pool[name] = {
+    std::make_unique<Tensor>(dim, false, Tensor::Initializer::NONE, name),
+    exec_order, lifespan, 0};
+
+  return pool[name].tensor.get();
+}
+
+/**
+ * @brief     Request tensor which has been already requested with the given
+ * spec
+ *
+ * @note returns empty tensor which will be filled when allocate is called.
+ * @note we assume that the caller checks if the exec_order and lifespan are
+ * compatible.
+ */
+Tensor *TensorPool::requestPreallocatedTensor(
+  const TensorDim dim, const std::vector<unsigned int> &exec_order,
+
+  TensorLifespan lifespan, const std::string &name) {
+  if (pool.find(name) == pool.end())
+    throw std::invalid_argument("Requested tensor not found");
+
+  auto &spec = pool[name];
+  if (spec.tensor->getDim() != dim)
+    throw std::invalid_argument("Request tensor dimension mismatch");
+
+  spec.exec_order.insert(spec.exec_order.end(), exec_order.begin(),
+                         exec_order.end());
+  spec.lifespan = enum_class_or<TensorLifespan>(spec.lifespan, lifespan);
+
+  return pool[name].tensor.get();
+}
+
+/**
+ * @brief finalize the requested tensors
+ *
+ * @details finalize the requested tensors, request memory for them and plan
+ * layout for their allocations.
+ */
+void TensorPool::finalize(const MemoryPlanner &planner,
+                          unsigned int start_order, unsigned int end_order) {
+  unsigned int bytes_requested = 0;
+  for (auto &entry : pool) {
+    auto &spec = entry.second;
+
+    if (spec.exec_order.empty())
+      continue;
+
+    /** 1. create the validity ranges for the all the requested tensors */
+    unsigned int validity_start =
+      *std::min_element(spec.exec_order.begin(), spec.exec_order.end());
+    unsigned int validity_end =
+      *std::max_element(spec.exec_order.begin(), spec.exec_order.end());
+
+    /**
+     * use lifespan to update the validity.
+     * if the validity is long term, the tensor must stay valid for the
+     * complete duration.
+     */
+    if (isTensorLongTerm(spec.lifespan)) {
+      validity_start = start_order;
+      validity_end = end_order;
+    }
+
+    /** 2. for each tensor request if it is in the provided range */
+    if (validity_end < start_order || validity_start > end_order)
```

Comment:
Isn't it error if the tensor validity is out of provided range?

Reply:
is the order inclusive?

Reply:
No, its not error. When the model is created, the validity of a tensor is given for training as well as for inference. However, when it is during `finalize()` that the mode of execution is actually finalized. So, in inference, many tensors validity will always be out of provided range.

Reply:
> is the order inclusive for both side?

yes - https://github.com/nnstreamer/nntrainer/blob/3d6138f9c4c839d3087a16cf5e6478bb947d5fe9/nntrainer/tensor/tensor_pool.h#L88-L89